zookeeper-server/src/main/java/org/apache/zookeeper/server/ServerCnxn.java

```diff
@@ -622,4 +622,21 @@ public String getHostAddress() {
     public String getSessionIdHex() {
         return "0x" + Long.toHexString(getSessionId());
     }
+
+    /**
+     * Disposes the ZooKeeperSaslServer instance, if any, and any
+     * system resources or security-sensitive information it may be
+     * using.  The field is also set to <code>null</code>.  This
+     * method is idempotent.
+     */
+    protected final void disposeSaslServer() {
+        if (zooKeeperSaslServer != null) {
```

Comment:
I am not sure about synchronisation around this variable.
It is the first time I am looking at its lifecycle and it seems to me that we are not doing well.
As you are setting it to **null** ...can we fall into an NPE?

Reply:
I had a closer look at the `ServerCnxn.close()` callers, and am now really perplexed.  Can that method be called from multiple/random threads?  Or are the calls otherwise somehow "serialized" wrt. to the memory model?  What is the easiest way to find out such things?

(Its implementations are not `synchronized`, but seem to be written in "naive" imperative style.  Am I missing something—or is that what you mean by "it seems to me that we are not doing well"? :)

In any case: that scared me a bit, so the soon to be pushed revised version:

 1. Does not `null` the field anymore, and just calls (idempotent) `dispose()`;
 
 2. Moves the `disposeSaslServer()` calls as far down the "leaves" of the `close()` implementations as possible, at a point where everything else has been unregistered and closed.

What do you think?  (Perhaps best to rely on GC finalizers, after all!)